### ISSUE TITLE
need Instructions for AWS credentials

### ISSUE THREAD

**Issue by nvadluri** [RoadieHQ/backstage-plugin-aws-auth#65]
Can you please help me out with the AWS credentials? 
1. Where I should update this? in which file ? or do I need to create new file and update this snippet

```
async function generateCredentials(backendUrl: string) {
  const reqBody = JSON.stringify({ RoleArn: 'arn:aws:iam::0123456789012:role/Example' });
  const resp = await (await fetch(`${backendUrl}/aws/credentials`, { body: reqBody })).json();
  return new AWS.Credentials({
    accessKeyId: resp.AccessKeyId,
    secretAccessKey: resp.SecretAccessKey,
    sessionToken: resp.SessionToken,
  });
}
AWS.config.credentials = await generateCredentials(backendUrl);
```

2. where I can store my AWS credentials? what file or any new file I need to create?

Clear instructions would be very helpful, as I am getting errors with my own steps.

**Comment by nvadluri** [comment]
Any help please?

**Comment by nvadluri** [comment]
<img width="553" alt="Screen Shot 2021-11-29 at 8 27 19 PM" src="https://user-images.githubusercontent.com/84935639/143969142-b18e2954-41f7-4fdb-9a6b-d6ed27117a7d.png">

I tried updating in many ways, but getting 404 error on the browser console logs and also seeing error message saying "Please add aws auth backend plugin"

**Comment by nvadluri** [comment]
Any help please?

**Comment by nvadluri** [comment]
<img width="1537" alt="Screen Shot 2021-12-02 at 2 24 10 PM" src="https://user-images.githubusercontent.com/84935639/144489350-97620aec-f1b2-40d4-a875-a0602b6bec3e.png">

**Comment by nvadluri** [comment]
Can anyone help please ? do I need to get the managed services to get help for this plug in ?  I have been waiting to get some instrcutinos on how to make this plugin workinig...

**Comment by martina-if** [comment]
Hi @nvadluri , sorry for the late reply here, we missed these messages.

This plugin was built some time ago and I was just going through the documentation and I agree with you that the instructions are not clear.

To be able to help you I'd like to understand what your use case is. Are you trying to use this together with the lambda plugin?

To try and unblock you I'll explain the way this normally works:

1. this aws-auth plugin is installed in the backend configured with long term AWS credentials (access key and secret)
2. a different frontend plugin (like the [aws-lambda](https://github.com/RoadieHQ/roadie-backstage-plugins/tree/main/plugins/frontend/backstage-plugin-aws-lambda)) that needs AWS credentials makes a request to the endpoint  `.../aws/credentials` to fetch temporary credentials
3. these temporary credentials are used in the frontend plugin to make AWS API calls (or use with the aws sdk)

The snippet you mentioned with the `generateCredentials(baseurl)` function is an example of step number 2.

If you are creating your own frontend plugin, that snippet is used before you use the AWS SDK.

If you are using the lambda plugin, this is already done for you and you don't need this snippet.

**Comment by nvadluri** [comment]
Thank you so much for the response. I am using roadie’s lambda plugin only (not my own)along with the backend auth plugin, followed all the steps as documented (for both aws lambda and backend plugins), I am getting the error install backed auth plugin and seeing on logs 404 for /api/aws/credentials. I deployed IAM user’s creds as env variables. When I tried to get lambda function details from the pod, I am getting the details. I don’t know what else i am missing if I can ignore that generate credentials code snippet. Please help me with the issue, I have been waiting for the response for a  week.

Thanks & Regards
Naveen Vadluri

________________________________

**Comment by nvadluri** [comment]
@martina-if and also I am seeing this on the logs
"2021-12-03T19:27:41.990Z aws warn AWS_ACCESS_KEY_ID and AWS_ACCESS_KEY_SECRET environment variables not set. Using default credentials provider chain. type=plugin"

**Comment by nvadluri** [comment]
@martina-if any solutions you found?

**Comment by martina-if** [comment]
Hi @nvadluri from that error I understand that the pod is not configured properly. Are you sure the environment variables `AWS_ACCESS_KEY_ID` and `AWS_ACCESS_KEY_ID` are set in the pod?

I'm not sure what you mean with this " I am getting the error install backed auth plugin". Do you have a few log lines you can post here?

My recommendation would be to get this working locally first.

**Comment by nvadluri** [comment]
Yes, I configured the pod with ENV variables, and I can get the AWS lambda function details from the pod.

root@backstage-5866c9c86c-dzz76:/app/kafka_2.13-3.0.0# aws lambda get-function --function-name test-ec2
{
    "Configuration": {
        "FunctionName": "test-ec2",
        "FunctionArn": "arn:aws:lambda:us-west-2:**********:function:test-ec2",
        "Runtime": "python3.8",
        "Role": "arn:aws:iam::********:role/service-role/test-ec2-role-rj02nj0b",
        "Handler": "lambda_function.lambda_handler",
        "CodeSize": 299,
        "Description": "",
        "Timeout": 3,
        "MemorySize": 128,
        "LastModified": "2021-07-09T13:57:04.788+0000",
        "CodeSha256": "**************=",
        "Version": "$LATEST",
        "TracingConfig": {
            "Mode": "PassThrough"
        },
        "RevisionId": "*****************"
    },
    "Code": {
        "RepositoryType": "S3",
        "Location": "https://awslambda-us-west-2-tasks.s3.us-west-2.amazonaws.com/snapshots/......
        ............"
    }
}
root@backstage-5866c9c86c-dzz76:/app/kafka_2.13-3.0.0#

about the error: "Please add aws auth backend plugin"
<img width="984" alt="Screen Shot 2021-12-06 at 11 17 37 AM" src="https://user-images.githubusercontent.com/84935639/144882213-b35baa50-f64e-420e-8fde-2c922ce08e89.png">

**Comment by nvadluri** [comment]
These are the related log chunks I found, related to lambda plugin

while the pod is starting up 
"

2021-12-06T16:21:09.382Z backstage info Loaded config from app-config.yaml
--
Mon, Dec 6 2021 11:21:09 am | 2021-12-06T16:21:09.394Z backstage info 3 secrets found in the config which will be redacted
Mon, Dec 6 2021 11:21:09 am | 2021-12-06T16:21:09.408Z backstage info Created UrlReader predicateMux{readers=azure{host=dev.azure.com,authed=false},bitbucket{host=bitbucket.org,authed=false},github{host=github.com,authed=true},gitlab{host=gitlab.com,authed=true},awsS3{host=amazonaws.com,authed=false},fetch{}
**Mon, Dec 6 2021 11:21:09 am | 2021-12-06T16:21:09.410Z aws warn AWS_ACCESS_KEY_ID and AWS_ACCESS_KEY_SECRET environment variables not set. Using default credentials provider chain. type=plugin**
Mon, Dec 6 2021 11:21:09 am | 2021-12-06T16:21:09.573Z auth info Configuring "database" as KeyStore provider type=plugin
Mon, Dec 6 2021 11:21:09 am | 2021-12-06T16:21:09.598Z techdocs info Creating Local publisher for TechDocs type=plugin

"

and also when load aws lambda component on the backstage

"

2021-12-06T16:25:28.491Z backstage info ::ffff:***** - - [06/Dec/2021:16:25:28 +0000] "GET /api/catalog/entities/by-name/component/default/lambda-rotate-keys/ancestry HTTP/1.1" 200 - "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Safari/537.36" type=incomingRequest
--
Mon, Dec 6 2021 11:25:28 am | 2021-12-06T16:25:28.493Z backstage info ::ffff:*******- - [06/Dec/2021:16:25:28 +0000] "GET /api/aws/credentials HTTP/1.1" 404 - "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.55 Safari/537.36" type=incomingRequest

"

**Comment by martina-if** [comment]
@nvadluri I think your pod is not getting the env variables as it's stated in the logs. The reason why the `aws lambda` command works could be that the cli is using the IMDS credentials or the machine credentials but the backstage plugin expects those variables, not any kind of AWS creds.

What happens when you run `kubectl exec -it <pod-name> -- env | grep AWS` ?

**Comment by nvadluri** [comment]
@martina-if 
I am seeing the keys which I applied as ENV variables. 
nvadluri@C02D72B7MD6T backstage-dcc % kubectl exec -it backstage-5866c9c86c-xtn8w -n digital-dcc-dev -- env | grep AWS
AWS_DEFAULT_REGION=us-west-2
AWS_ACCESS_KEY_ID=Myaccesskey
AWS_SECRET_ACCESS_KEY=mysecuritykey
nvadluri@C02D72B7MD6T backstage-dcc %

root@backstage-5866c9c86c-xtn8w:/app# aws sts get-caller-identity
{
    "UserId": "MYnewIAMUSER",
    "Account": "123456",
    "Arn": "arn:aws:iam::1234556:user/dcc-backstage-user"
}
root@backstage-5866c9c86c-xtn8w:/app#

Just fyi-
This POD is running in a different aws account and the lambda function is in another account. I created a new IAM user in the account where the lambda function is and with a policy attached (w/ lambda get-function) and also role with STS: assume. I got the credentials for that IAM user and deployed as ENV varaibales and I am seeing that user only (posted the details above)

**Comment by martina-if** [comment]
I think the issue is that [the code](https://github.com/RoadieHQ/backstage-plugin-aws-auth/blob/main/src/service/router.ts#L27) expects `AWS_ACCESS_KEY_SECRET` and your environment variable is called `AWS_SECRET_ACCESS_KEY`. Can you try to rename it?

**Comment by nvadluri** [comment]
@martina-if 
I have updated as you said, and this is my deployment file chunk
"          envFrom:
            - secretRef:
                name: postgres-secrets
            - secretRef:
                name: backstage-secrets
            - secretRef:
                name: backstage-gitlab-secrets
          env:
            - name: AWS_ACCESS_KEY_ID
              valueFrom: 
                secretKeyRef:
                  name: aws-secrets
                  key: AWS_ACCESS_KEY_ID
            - name: AWS_ACCESS_KEY_SECRET
              valueFrom: 
                secretKeyRef:
                  name: aws-secrets
                  key: AWS_ACCESS_KEY_SECRET
            - name: AWS_DEFAULT_REGION
              valueFrom: 
                secretKeyRef:
                  name: aws-secrets
                  key: AWS_DEFAULT_REGION
            - name: LOG_LEVEL
              value: debug"
              
     and this is how I set the secrets:
     
apiVersion: v1
kind: Secret
metadata:
  name: aws-secrets
  namespace: digital-dcc-dev
type: Opaque
data:
  AWS_ACCESS_KEY_ID: xxxxxxxxxx=
  AWS_ACCESS_KEY_SECRET: xxxxxxxx==
  AWS_DEFAULT_REGION: xxxxxx
     
  Now I am NOT seeing in the startup logs saying 
  "aws warn AWS_ACCESS_KEY_ID and AWS_ACCESS_KEY_SECRET environment variables not set"
  but seeing when I load the component 
   "2021-12-06T17:53:50.786Z backstage info ::ffff:10.0.0.0 - - [06/Dec/2021:17:53:50 +0000] "GET /api/aws/credentials HTTP/1.1" 404" 
   
   and also on the backstage screen the error "Please add aws auth backend plugin" still showing.

**Comment by nvadluri** [comment]
@martina-if 
Hurrayyyy...it worked finally...along with your suggestion, I did revert one change which I did before your response, to fix the issue on my own, it was still there, so reverted back now...(  **FROM**   apiRouter.use('/aws/credentials', **TO** await aws(awsEnv));apiRouter.use('/aws', await aws(awsEnv)); 

and also I have added  "AWS_SECRET_ACCESS_KEY" along with " AWS_ACCESS_KEY_SECRET" env variable..


apiVersion: v1
kind: Secret
metadata:
  name: aws-secrets
  namespace: digital-dcc-dev
type: Opaque
data:
  AWS_ACCESS_KEY_ID: **********=
  AWS_SECRET_ACCESS_KEY: **********==
  AWS_ACCESS_KEY_SECRET: **********==
  AWS_DEFAULT_REGION: ********
  
  Thank you so much for finding the issue...You are awesome!

**Comment by martina-if** [comment]
Glad it's working! You're welcome! I will close this issue then :)
